src/hugr/rewrite/replace.rs

```diff
@@ -0,0 +1,591 @@
+    NoRemovedEdge(NewEdgeSpec),
+    /// The [NewEdgeKind] was not applicable for the source/target node(s)
+    #[error("The edge kind was not applicable to the {0:?} node: {1:?}")]
+    BadEdgeKind(Direction, NewEdgeSpec),
+}
+
+/// A Hugr or portion thereof that is part of the [Replacement]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum WhichHugr {
+    /// The newly-inserted nodes, i.e. the [Replacement::replacement]
+    Replacement,
+    /// Nodes in the existing Hugr that are not [Replacement::removal]
+    /// (or are on the RHS of an entry in [Replacement::transfers])
+    Retained,
+}
+
+impl std::fmt::Display for WhichHugr {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(match self {
+            Self::Replacement => "replacement Hugr",
+            Self::Retained => "retained portion of Hugr",
+        })
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::collections::HashMap;
+
+    use crate::builder::{BuildError, CFGBuilder, Container, Dataflow, HugrBuilder};
+    use crate::extension::prelude::USIZE_T;
+    use crate::extension::{ExtensionRegistry, ExtensionSet, PRELUDE};
+    use crate::hugr::{HugrMut, NodeType};
+    use crate::ops::handle::{BasicBlockID, ConstID, NodeHandle};
+    use crate::ops::{self, BasicBlock, LeafOp, OpTrait, OpType, DFG};
+    use crate::std_extensions::collections;
+    use crate::types::{FunctionType, Type, TypeArg, TypeRow};
+    use crate::{type_row, Hugr, HugrView, Node, OutgoingPort};
+
+    use super::{NewEdgeKind, NewEdgeSpec, Replacement};
+
+    #[test]
+    fn cfg() -> Result<(), Box<dyn std::error::Error>> {
+        let reg: ExtensionRegistry = [PRELUDE.to_owned(), collections::EXTENSION.to_owned()].into();
+        let listy = Type::new_extension(
+            collections::EXTENSION
+                .get_type(collections::LIST_TYPENAME.as_str())
+                .unwrap()
+                .instantiate_concrete([TypeArg::Type { ty: USIZE_T }])
+                .unwrap(),
+        );
+        let pop: LeafOp = collections::EXTENSION
+            .instantiate_extension_op("pop", [TypeArg::Type { ty: USIZE_T }], &reg)
+            .unwrap()
+            .into();
+        let push: LeafOp = collections::EXTENSION
+            .instantiate_extension_op("push", [TypeArg::Type { ty: USIZE_T }], &reg)
+            .unwrap()
+            .into();
+        let just_list = TypeRow::from(vec![listy.clone()]);
+        let exset = ExtensionSet::singleton(&collections::EXTENSION_NAME);
+        let intermed = TypeRow::from(vec![listy.clone(), USIZE_T]);
+
+        let mut cfg = CFGBuilder::new(
+            FunctionType::new_linear(just_list.clone()).with_extension_delta(&exset),
+        )?;
+
+        let pred_const = cfg.add_constant(
+            ops::Const::unary_unit_sum(),
+            // Slightly awkward we have to specify this, and the choice is non-obvious:
```

Comment:
I felt I had to say something here, as this relates to #633....
* at the moment Builder `add_constant` forces you to specify the Extensions. The obvious choice would be `ExtensionSet::new()` but then you would need lift nodes, ugh
* If we generalized `add_constant` to allow specifying "open" (None), then in theory inference should figure out the choice here, but in practice #633 kicks in and it only does so *sometimes* - infinite-looping on other runs :(

So without generalizing add_constant (to no immediate win here), I don't really see what I can say, so I will try to come back to this after fixing #633....